psytran/loop.py

```diff
@@ -216,3 +216,36 @@ def is_parallelisable(loop):
     :rtype: :py:class:`bool`
     """
     return loop.independent_iterations()
+
+
+def get_perfectly_nested_loops(schedule):
+    """
+    Finds the outermost of all perfectly nested loop structures within a
+    supplied schedule and returns them as a list.
+
+    Does NOT nexcessarily return an outer loop, only the topmost loop of a
+    perfectly nested structure.
+
+    :arg schedule: the schedule to search.
+    :type schedule: :py:class:`Schedule`
+
+    :returns loops: all loops that are the outermost of a perfectly nested
+                    structure.
+    :rtype loops: list[:py:class:`Loop`]
+    """
+    loops = []
+
+    # Test all loops in schedule to see if they are perfectly nested
+    for loop in schedule.walk(nodes.Loop):
+        if is_perfectly_nested(loop):
+            loops.append(loop)
+
+    # For each loop in the list check if loops below it are a descendent and
+    # remove them if they are
+    for top_loop in loops[:]:
+        for bottom_loop in reversed(loops[:]):
+            if bottom_loop in get_descendents(top_loop):
+                loops.remove(bottom_loop)
```

Comment:
I considered suggesting the following refactor but I'm not sure it's actually any more efficient. Depends on the relative sizes of `loops` vs descendents.
```suggestion
    for top_loop in loops[:]:
        for bottom_loop in get_descendents(top_loop):
            if bottom_loop in loops:
                loops.remove(bottom_loop)
```

Reply:
Hmm I think yours is more efficient since you only calculate the descendants for each loop once whereas mine calculates the descendants of the top loop on each item in the second loop.